src/openms/source/ANALYSIS/ID/MetaboliteSpectralMatching.cpp

```diff
@@ -410,6 +411,138 @@ namespace OpenMS
   }
 
 
+double log_factorial_(int n) {
+    double result = 0.0;
+    for (int i = 2; i <= n; ++i) {
+        result += std::log(i);
+    }
+    return result;
+}
+
+// Log survival function (log-SF) for Poisson distribution
+double log_poisson_sf_(int x, double lambda) {
```

Comment:
I see you are already using boost:
maybe consider something like:
```
    poisson_distribution<> poisson(lambda);

    // Log survival function using complement
    double sf = cdf(complement(poisson, x));
    return std::log(sf);
```

Reply:
I tried using the standard CDF function, but the values are so large that I kept getting underflow issues.